File: PROG_길찾기게임_200508.cpp
/* 길 찾기 게임 Level 3
https://programmers.co.kr/learn/courses/30/lessons/42892
 */
#include <algorithm>
#include <vector>
using namespace std;

class Node{
public:
    int idx, x, y;
    int left=-1, right=-1;
};

bool cmpX(const Node& n1, const Node& n2){
    return n1.x < n2.x;
}

bool cmpY(const Node& n1, const Node& n2){
    return n1.y < n2.y;
}

int make_tree(vector<Node>& tree, int start, int end){
    if(start == end) return -1;
    if(end-start <= 1) return start;
    
    int mid = max_element(tree.begin()+start, tree.begin()+end, cmpY) - tree.begin();
    tree[mid].left = make_tree(tree, start, mid);
    tree[mid].right = make_tree(tree, mid+1, end);
    return mid;
}

void preorder(vector<Node>& tree, vector<int>& list, int now){
    list.push_back(tree[now].idx);
    if(tree[now].left != -1) preorder(tree, list, tree[now].left);
    if(tree[now].right != -1) preorder(tree, list, tree[now].right);
}

void postorder(vector<Node>& tree, vector<int>& list, int now){
    if(tree[now].left != -1) postorder(tree, list, tree[now].left);
    if(tree[now].right != -1) postorder(tree, list, tree[now].right);
    list.push_back(tree[now].idx);
}

vector<vector<int>> solution(vector<vector<int>> nodeinfo) {
    vector<Node> tree(nodeinfo.size());
    for (int i=0; i<nodeinfo.size(); i++) {
        tree[i].idx = i+1;
        tree[i].x = nodeinfo[i][0];
        tree[i].y = nodeinfo[i][1];
    }
    sort(tree.begin(), tree.end(), cmpX);

    int top = make_tree(tree, 0, tree.size());

    vector<vector<int>> answer(2);
    answer[0].reserve(nodeinfo.size());
    answer[1].reserve(nodeinfo.size());
    preorder(tree, answer[0], top);
    postorder(tree, answer[1], top);
    
    return answer;
}

File: Day29/PROG_길찾기게임_200508.java
/*
 * programmers lv3 길 찾기 게임
 */

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;

public class PROG_길찾기게임_200508 {
	public static void main(String[] args) {
		int[][] answer = solution(new int[][] {{5,3},{11,5},{13,3},{3,5},{6,1},{1,3},{8,6},{7,2},{2,2}});
		for (int i = 0; i < answer.length; i++) {
			System.out.println(Arrays.toString(answer[i]));
		}
	}
	
	public static int[][] solution(int[][] nodeinfo) {
		int[][] answer = new int[2][nodeinfo.length]; // 전위 후위 두 가지 방법
		
		ArrayList<Node> list = new ArrayList<>();
		for (int i = 0; i < nodeinfo.length; i++) {
			list.add(new Node(i+1, nodeinfo[i][0], nodeinfo[i][1]));
		}
		
		Collections.sort(list);
//		System.out.println(list.toString());
		
		Node root = list.get(0);
		for (int i = 1; i < list.size(); i++) {
			addNode(root, list.get(i));
		}
		
		ArrayList<Integer> tmp = new ArrayList<>(); 
		preorder(root, tmp);
		for (int i = 0; i < answer[0].length; i++) {
			answer[0][i] = tmp.get(i);
		}
		
		tmp.clear();
		
		postorder(root, tmp);
		for (int i = 0; i < answer[1].length; i++) {
			answer[1][i] = tmp.get(i);
		}
		
		return answer;
	}
	
	private static void postorder(Node node, ArrayList<Integer> tmp) {
		if(node == null)
			return;
		
		postorder(node.left, tmp);
		postorder(node.right, tmp);
		tmp.add(node.idx);
	}

	private static void preorder(Node node, ArrayList<Integer> tmp) {
		if(node == null)
			return;
		
		tmp.add(node.idx);
		preorder(node.left, tmp);
		preorder(node.right, tmp);
		
	}

	private static void addNode(Node parent, Node child) {
		if(child.x < parent.x) {
			if(parent.left == null)
				parent.left = child;
			else 
				addNode(parent.left, child);
		} else {
			if(parent.right == null)
				parent.right = child;
			else
				addNode(parent.right, child);
		}
		
	}

	static class Node implements Comparable<Node>{
		int idx, x, y;
		Node left = null;
		Node right = null;
		public Node(int idx, int x, int y) {
			super();
			this.idx = idx;
			this.x = x;
			this.y = y;
		}

		@Override
		public String toString() {
			return "Node [idx=" + idx + ", x=" + x + ", y=" + y + "]";
		}

		@Override
		public int compareTo(Node o) {
			if(this.y == o.y)
				return Integer.compare(this.x, o.x);
			else
				return -Integer.compare(this.y, o.y);
		}
	}
}
